### PR TITLE
[libc][CPP] Add all_of and find_if_not to algorithm.h

### PR DESCRIPTION
**Pull request by PiJoules**
This is needed for the allocator implementation for malloc.

### CODE REVIEW

File: libc/test/src/__support/CPP/algorithm_test.cpp [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "src/__support/CPP/algorithm.h"
+#include "src/__support/CPP/array.h"
+#include "test/UnitTest/Test.h"
+
+namespace LIBC_NAMESPACE::cpp {
+
```
Comment (michaelrj-google) [inline libc/test/src/__support/CPP/algorithm_test.cpp:14]: nit: add a TODO and file an issue to add unit tests for the rest of `CPP/algorithm`

**Review by michaelrj-google — APPROVED** [review]
Overall LGTM

File: libc/test/src/__support/CPP/algorithm_test.cpp (hunk 1 above)
Comment (PiJoules) [reply libc/test/src/__support/CPP/algorithm_test.cpp:16]: Done. https://github.com/llvm/llvm-project/issues/94066

**Comment by michaelrj-google** [comment]
This patch is fine to land, but I'd recommend waiting for the presubmits to finish and submitting it next week.

**Review by lntue — APPROVED** [review]

**Pull request merged**